Allow user to assert no mask contiguous check is necessary

mikekgfb opened the pull request:
Summary:
Allow user to assert no mask contiguous check is necessary:
(1) Prevents sync event which will disrupt CUDA Graph collection, and
(2) offers slightly better performance by avoid a sync

This needs to be a separate opt-in option because we change behavior of malformed masks.  It's the only way to get BT into CUDA Graph based on what I understood about CUDA Graph collection from ngimel.

Test Plan: sandcastle unit tests

Differential Revision: D38040418

jbschlosser reviewed (commented):
From our offline discussion, I understand that:
* The flag name `mask_contiguous_check` is planned to change to avoid confusion with the standard PyTorch definition of "contiguous"
* A set of unit tests is being added to ensure that the mask correctly affects the logic

FYI on the OSS side, we'd typically require both of those in this PR before landing. This keeps PRs self-contained and helps minimizes nightly breakage. Would it be a huge pain to stick to this policy here?

Commit:
Allow user to assert no mask contiguous check is necessary (#82533)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/82533

Allow user to assert no mask contiguous check is necessary:
(1) Prevents sync event which will disrupt CUDA Graph collection, and
(2) offers slightly better performance by avoid a sync

This needs to be a separate opt-in option because we change behavior of malformed masks.  It's the only way to get BT into CUDA Graph based on what I understood about CUDA Graph collection from ngimel.

Test Plan: sandcastle unit tests

Differential Revision: D38040418

fbshipit-source-id: 77e44086d0000d8de965cac51580269cf420f2c8

jbschlosser commented on aten/src/ATen/native/nested/NestedTensorMath.cpp:
```diff
@@ -165,7 +165,8 @@ Tensor NestedTensor_nested_tensor_from_mask(const Tensor& t, const Tensor& mask)
     nums = nums.to(sizes.options());
 
-    TORCH_CHECK(sizes.equal(nums), "Mask must be left-aligned without gaps");
+    if (mask_check)
+      TORCH_CHECK(sizes.equal(nums), "Mask must be left-aligned without gaps");
```
Seems like `nums` is only used for this check. Should we avoid computing it at all when `mask_check=True`?

jbschlosser reviewed (approved):
Stamping, but expecting immediate follow-ups with unit tests and optimization

zrphercule reviewed (approved):
This diff LGTM, please make sure CI is also happy.

mikekgfb commented:
topic: new feature 
release notes: better transformer
